src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD201CancelAfterSwitchToMainThreadAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+                    var cancellationTokenNoneSymbol = cancellationTokenTypeSymbol.GetMembers(nameof(CancellationToken.None)).Single();
+                    startCtxt.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, switchMethods, cancellationTokenTypeSymbol, cancellationTokenNoneSymbol)), OperationKind.Invocation);
+                }
+            });
+        }
+
+        private void AnalyzeInvocation(OperationAnalysisContext context, ImmutableArray<ISymbol> switchMethods, INamedTypeSymbol cancellationTokenTypeSymbol, ISymbol cancellationTokenNoneSymbol)
+        {
+            var invocationOperation = (IInvocationOperation)context.Operation;
+            if (invocationOperation.TargetMethod is object && switchMethods.Contains(invocationOperation.TargetMethod))
+            {
+                IArgumentOperation? tokenArgument = invocationOperation.Arguments.FirstOrDefault(arg => arg.Value?.Type?.Equals(cancellationTokenTypeSymbol) ?? false);
+
+                // Filter out missing arguments, `default` and `default(CancellationToken)`
+                if (tokenArgument is null || tokenArgument.IsImplicit || tokenArgument.Value.IsImplicit || tokenArgument.Value is IDefaultValueOperation)
+                {
+                    return;
+                }
+
+                // Did CancellationToken.None get passed in?
+                if (tokenArgument.Value is IPropertyReferenceOperation p && p.Member.Equals(cancellationTokenNoneSymbol))
+                {
+                    return;
+                }
+
+                // Did the invocation get followed by a check on that CancellationToken?
+                var statement = Utils.FindAncestor<IExpressionStatementOperation>(invocationOperation);
+                if (statement?.Parent is IBlockOperation containingBlock)
+                {
+                    int currentStatement = containingBlock.Operations.IndexOf(statement);
+                    IOperation? nextOperation = containingBlock.Operations.Length > currentStatement + 1 ? containingBlock.Operations[currentStatement + 1] : null;
+                    if (!IsTokenCheck(nextOperation, tokenArgument.Value))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, invocationOperation.Syntax.GetLocation()));
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, invocationOperation.Syntax.GetLocation()));
+                }
+
+                bool IsTokenCheck(IOperation? consideredStatement, IOperation token)
+                {
+                    // if (token.IsCancellationRequested)
+                    if (consideredStatement is IConditionalOperation ifStatement &&
+                        ifStatement.Condition is IMemberReferenceOperation memberAccess &&
+                        Utils.IsSameSymbol(memberAccess.Instance, token) &&
```

Comment:
💡 Would be good to have a test like this:

```
async Task Method(JoinableTaskFactory jtf, Context first, Context second)
{
  await jtf.SwitchToMainThreadAsync(first.Token);
  second.Token.ThrowIfCancellationRequested(); // VSTHRD201: same property but different instances
}
```

Reply:
@CyrusNajmabadi mentioned on gitter.im today the same concern, and I'm pretty sure that test would fail. I think that's an acceptably low risk though for the time being. When we get to data flow analysis in this library where we can track values it might be interesting to revisit.

Reply:
I went ahead and added the test. As predicted, it fails.